utils/wasm-gen/src/config/syscalls/param.rs

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Entities describing sys-call param, more precisely, it's allowed values.
+//!
+//! Types here are used to create [`crate::SysCallsConfig`].
+
+use arbitrary::{Result, Unstructured};
+use gear_wasm_instrument::{
+    parity_wasm::elements::ValueType,
+    syscalls::{ParamType, SysCallName},
+};
+use std::{collections::BTreeMap, ops::RangeInclusive};
+
+#[derive(Debug)]
+pub(crate) enum ProcessedSysCallParams {
+    Alloc,
+    Value {
+        value_type: ValueType,
+        allowed_values: Option<SysCallParamAllowedValues>,
+    },
+    MemoryArray,
+    MemoryPtrValue,
+}
+
+pub(crate) fn process_sys_call_params(
+    name: Option<SysCallName>,
+    params: &[ParamType],
+    params_config: &SysCallsParamsConfig,
+) -> Vec<ProcessedSysCallParams> {
+    let skip_mem_array = name
+        .map(|name| matches!(name, SysCallName::SendInput | SysCallName::SendInputWGas))
+        .unwrap_or_default();
+
+    let mut res = Vec::with_capacity(params.len());
+    for &parameter in params {
+        if ParamType::Size == parameter {
+            if let Some((ParamType::Ptr, is_memory_array @ false)) = res.last_mut() {
+                if !skip_mem_array {
+                    *is_memory_array = true;
+                    continue;
+                }
+            }
+        }
+        res.push((parameter, false));
+    }
+
+    res.into_iter()
+        .map(|(param_type, is_memory_array)| match is_memory_array {
+            true => ProcessedSysCallParams::MemoryArray,
+            false => match param_type {
+                ParamType::Alloc => ProcessedSysCallParams::Alloc,
+                ParamType::Ptr => ProcessedSysCallParams::MemoryPtrValue,
+                _ => ProcessedSysCallParams::Value {
+                    value_type: param_type.into(),
+                    allowed_values: params_config.get_rule(&param_type),
+                },
+            },
+        })
+        .collect()
+}
+
+/// Sys-calls params config.
+///
+/// This is basically a map, which creates a relationship between each kind of
+/// param, that a sys-call can have, and allowed values ("rules") for each of
+/// the params.
+#[derive(Debug, Clone)]
+pub struct SysCallsParamsConfig(BTreeMap<ParamType, SysCallParamAllowedValues>);
```

Comment:
Is `BtreeMap` better that `HashMap`?

Reply:
well, yeah, `HashMap` is better, as we only read data from config, which is O(1).

Reply:
done